### PR TITLE
Initial stream implementation

### PR DESCRIPTION
**Pull request by kleewho**

### CODE REVIEW

File: src/main/kotlin/com/pubnub/api/crypto/data/SizedInputStream.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package com.pubnub.api.crypto.data
+
+import java.io.InputStream
+
+class SizedInputStream(
+    private val inputStream: InputStream,
+    val size: Int
```
Comment (marcin-cebo) [inline src/main/kotlin/com/pubnub/api/crypto/data/SizedInputStream.kt:7]: I don't see "size" being used anywhere in this class. 
Is it used?

File: src/main/kotlin/com/pubnub/api/crypto/cryptor/AesCbcCryptor.kt [hunk 2]
```diff
@@ -55,12 +58,40 @@ class AesCbcCryptor(val cipherKey: String) : Cryptor {
...
+            val ivSpec: AlgorithmParameterSpec = IvParameterSpec(ivBytes)
+            val keyBytes: ByteArray = getKeyAsShaInHexBytes(cipherKey)
+            val newKey = SecretKeySpec(keyBytes, "AES")
+            val cipher = Cipher.getInstance(CIPHER_TRANSFORMATION)
+            cipher.init(Cipher.ENCRYPT_MODE, newKey, ivSpec)
+            val cipheredStream = CipherInputStream(stream, cipher)
+            val cipheredBytes = cipheredStream.readBytes()
+
+            return EncryptedStreamData(
+                stream = SizedInputStream(
```
Comment (marcin-cebo) [inline src/main/kotlin/com/pubnub/api/crypto/cryptor/AesCbcCryptor.kt:73]: I see that metadata from EncryptedStreamData are not use here, we don't store IV there  as in EncryptedData ?

File: src/main/kotlin/com/pubnub/api/crypto/data/SizedInputStream.kt (hunk 1 above)
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/crypto/data/SizedInputStream.kt:7]: It is not. Size is not needed here but it's going to be used elsewhere

File: src/main/kotlin/com/pubnub/api/crypto/cryptor/AesCbcCryptor.kt (hunk 2 above)
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/crypto/cryptor/AesCbcCryptor.kt:73]: It's should be there I think

File: src/main/kotlin/com/pubnub/api/crypto/data/SizedInputStream.kt (hunk 1 above)
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/crypto/data/SizedInputStream.kt:7]: You know what. It's not important. Either way we will have to materialize this stream eventually if sending it out with uploadfile, but we can do it there, so to make the code simpler I will move computing the size to there

File: src/main/kotlin/com/pubnub/api/crypto/cryptor/LegacyCryptor.kt [hunk 3]
```diff
@@ -57,28 +62,57 @@ class LegacyCryptor(val cipherKey: String, val useRandomIv: Boolean = true) : Cr
...
+        }
     }
 
     private fun getKeyAsShaInHexBytes(cipherKey: String): ByteArray {
-        val cipherKeyBytes = cipherKey.toByteArray(Charsets.UTF_8)
-        val cipherKeySha256Bytes: ByteArray = MessageDigest.getInstance("SHA-256").digest(cipherKeyBytes)
-        val hexString = cipherKeySha256Bytes.joinToString("") { "%02x".format(it) }
-        return hexString.toByteArray(Charsets.UTF_8)
+        return String(
+            Crypto.hexEncode(Crypto.sha256(cipherKey.toByteArray(charset(ENCODING_UTF_8)))),
```
Comment (marcin-cebo) [inline src/main/kotlin/com/pubnub/api/crypto/cryptor/LegacyCryptor.kt:104]: I assume that this is temporary solution and we want to replace Crypto.java with equivalent code in kotlin?

Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/crypto/cryptor/LegacyCryptor.kt]: I move the methods to cryptors

File: src/main/kotlin/com/pubnub/api/PubNub.kt [hunk 4]
```diff
@@ -1770,6 +1776,12 @@ class PubNub internal constructor(
         shouldStore: Boolean? = null,
         cipherKey: String? = null
     ): SendFile {
+        val cryptoModule = if (cipherKey != null) {
+            CryptoModule.createLegacyCryptoModule(cipherKey)
```
[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/PubNub.kt:1780]: Yes. This is to preserve functionality that we had previously

File: src/main/kotlin/com/pubnub/api/PubNub.kt [hunk 5]
```diff
@@ -1997,8 +2015,14 @@ class PubNub internal constructor(
...
-    ): InputStream = encrypt(inputStream, cipherKey)
+        cipherKey: String? = null
+    ): InputStream = getCryptoModuleOrThrow(cipherKey).encryptStream(inputStream)
+
+    private fun getCryptoModuleOrThrow(cipherKey: String? = null): CryptoModule {
+        return cipherKey?.let {
+            CryptoModule.createLegacyCryptoModule(it, configuration.useRandomInitializationVector)
+        } ?: configuration.cryptoModule ?: throw PubNubException("Crypto module is not initialized")
+    }
     //endregion
```
[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/api/PubNub.kt:2026]: Yes. This is to preserve functionality that we had previously

File: src/main/kotlin/com/pubnub/extension/CryptoModule.kt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package com.pubnub.extension
```
[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/extension/CryptoModule.kt:1]: I want the crypto module to have as little dependencies as possible so I didn't put it there

File: src/main/kotlin/com/pubnub/api/PubNub.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/api/PubNub.kt:1780]: ok

File: src/main/kotlin/com/pubnub/extension/CryptoModule.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/kotlin/com/pubnub/extension/CryptoModule.kt:1]: ok

[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/kotlin/com/pubnub/extension/CryptoModule.kt]: Moved

**Pull request merged**